serving/src/main/java/feast/serving/encoding/FeatureRowDecoder.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright 2018-2020 The Feast Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package feast.serving.encoding;
+
+import feast.core.FeatureSetProto.FeatureSetSpec;
+import feast.core.FeatureSetProto.FeatureSpec;
+import feast.types.FeatureRowProto.FeatureRow;
+import feast.types.FieldProto.Field;
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class FeatureRowDecoder {
+
+  private final String featureSetRef;
+  private final FeatureSetSpec spec;
+
+  public FeatureRowDecoder(String featureSetRef, FeatureSetSpec spec) {
+    this.featureSetRef = featureSetRef;
+    this.spec = spec;
+  }
+
+  /**
+   * A feature row is considered encoded if the feature set and field names are not set. This method
+   * is required for backward compatibility purposes, to allow Feast serving to continue serving non
+   * encoded Feature Row ingested by an older version of Feast.
+   *
+   * @param featureRow Feature row
+   * @return boolean
+   */
+  public Boolean isEncoded(FeatureRow featureRow) {
+    return featureRow.getFeatureSet().isEmpty()
+        && featureRow.getFieldsList().stream().allMatch(field -> field.getName().isEmpty());
+  }
+
+  /**
+   * Validates if an encoded feature row can be decoded without exception.
+   *
+   * @param featureRow Feature row
+   * @return boolean
+   */
+  public Boolean isEncodingValid(FeatureRow featureRow) {
+    return featureRow.getFieldsList().size() == spec.getFeaturesList().size();
```

Comment:
If you discover an old feature row (that is still within max age), how will you interpret it? Are we using the versions here to be able to interpret old feature rows?

Reply:
Redis Key contains both the feature set name and version. Since both ingestion and serving encode/decode based on the feature set name and version specified in Redis Key, the spec should be consistent, and therefore isEncodingValid will always return true.

Originally i added this method to partially address the concern raised in #515:

> If field values are only going to be associated with field names at runtime by external configuration has any thought been given to a method for ensuring that the same configuration that was used to write the data is the configuration used to read the data? Something such as a checksum/fingerprint of the FeatureSet configuration stored alongside the data in Redis (or in the key) will help prevent a mismatch of configuration and data due to a bug somewhere else in the system.

Perhaps i should remove this method from this PR?

Reply:
For this thread and the PR as a whole especially the part quoted below, does this approach fall down and need to be reworked as versions get removed?

> In cases where the Feature Row are from external Kafka source, where it is possible for the feature row to be malformed, there are a few checks in place:
> 1. It's fine for the feature row to have feature fields in a different order than that specified in the feature set spec, as the encoding process will sort the feature in alphabetical order based on name.

Reply:
Oh yeah, @zhilingc has brought this up [in a comment thread on her versions removal RFC](https://docs.google.com/document/d/1P44LHd724JloQtpn10naAg5MlzcrkeOlN3QzAn9O8uY/edit?disco=AAAAJReHWQM). Ouch 🤕